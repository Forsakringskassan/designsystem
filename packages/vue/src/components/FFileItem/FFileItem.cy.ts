import { type DefineComponent, defineComponent } from "vue";
import { FFileItemPageObject } from "../../cypress";
import { FIcon } from "../FIcon";
import { FProgressbar } from "../FProgressbar";
import FFileItem from "./FFileItem.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        components: { FFileItem, FIcon, FProgressbar },
        data() {
            return {
                fileName: "bar.pdf",
                mimeType: "application/pdf",
                progress: 30,
                filteredProgress: 30,
            };
        },
    });
}

describe("FFileItem", () => {
    const fileItem = new FFileItemPageObject("[data-test=file-item]");

    it("Should have a page object that can access any necessary elements", () => {
        const template = /* HTML */ `
            <div>
                <label>
                    Progress:
                    <input
                        id="upload-progress"
                        v-model.number="progress"
                        type="number"
                        style="margin-bottom: 20px"
                        @input="filteredProgress = progress === '' ? 0 : progress"
                /></label>
                <f-file-item
                    v-test="'file-item'"
                    :file-name="fileName"
                    :mime-type="mimeType"
                >
                    <template #row>
                        <button
                            v-if="progress < 100"
                            type="button"
                            class="button button--discrete file-item__file-remove file-item__abort"
                        >
                            <f-icon name="close" class="button__icon"></f-icon>
                            <span> Avbryt uppladdning </span>
                        </button>
                        <button
                            v-else-if="progress === 100"
                            type="button"
                            class="button button--discrete file-item__file-remove"
                        >
                            <f-icon
                                name="trashcan"
                                class="button__icon"
                            ></f-icon>
                            Ta bort
                        </button>
                    </template>
                    <f-progressbar
                        v-if="progress < 100"
                        aria-label="progress"
                        :value="filteredProgress"
                    ></f-progressbar>
                </f-file-item>
            </div>
        `;
        cy.mount(createComponent(template));
        fileItem.fileName().should("have.trimmedText", "bar.pdf");
        fileItem.typeOfFile().should("be.equal", "pdf");
        fileItem
            ._cancelDeleteButton()
            .should("have.trimmedText", "Avbryt uppladdning");
        fileItem.typeOfButtonIcon().should("be.equal", "close");
        const selector = "#upload-progress";
        cy.get(selector).clear();
        cy.get(selector).type("30");
        fileItem._progressMeter.progressMeter().should("be.visible");
        cy.get(selector).clear();
        cy.get(selector).type("100");
        fileItem._progressMeter.progressMeter().should("not.exist");
        fileItem.typeOfButtonIcon().should("be.equal", "trashcan");
    });
});
