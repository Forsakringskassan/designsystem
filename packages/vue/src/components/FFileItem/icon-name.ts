/**
 * @public
 */
export type IconName = "doc" | "file" | "pic" | "pdf";
