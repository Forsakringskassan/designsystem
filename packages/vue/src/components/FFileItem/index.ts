export { default as FFileItem } from "./FFileItem.vue";
export { type IconName as FFileItemIconName } from "./icon-name";
