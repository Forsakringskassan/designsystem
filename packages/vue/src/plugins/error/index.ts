export {
    type ErrorPluginOptions,
    default as ErrorPlugin,
    UNHANDLED_ERROR_EVENT,
} from "./ErrorPlugin";
export { default as FErrorHandlingApp } from "./FErrorHandlingApp.vue";
