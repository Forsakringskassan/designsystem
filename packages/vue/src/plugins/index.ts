export * from "./test";
export * from "./translation";
export * from "./validation";
export * from "./error";
export { FormatPlugin } from "./format";
