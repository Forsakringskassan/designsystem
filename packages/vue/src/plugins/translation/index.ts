/* istanbul ignore file */
export { type TranslateFunction } from "./translate-function";
export { TranslationMixin, TranslationPlugin } from "./TranslationPlugin";
export { useTranslate } from "./use-translate";
