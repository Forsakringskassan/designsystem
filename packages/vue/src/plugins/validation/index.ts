/* istanbul ignore file */
export { ValidationPlugin } from "./ValidationPlugin";
