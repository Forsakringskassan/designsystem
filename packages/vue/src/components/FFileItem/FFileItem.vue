<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService, isSet } from "@fkui/logic";
import FIcon from "../FIcon/FIcon.vue";
import { type IconName } from "./icon-name";

/* default icon used when no mimetype matches */
const DEFAULT_ICON: IconName = "file";

/* map mimetype -> icon */
const iconMap: Record<string, IconName> = {
    "image/*": "pic",
    "application/msword": "doc",
    "application/pdf": "pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "doc",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.template": "doc",
};

export default defineComponent({
    name: "FFileItem",
    components: { FIcon },
    inheritAttrs: false,
    props: {
        /**
         * The id for the input id attribute.
         * The id for the label for attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * The file name.
         */
        fileName: {
            type: String,
            required: true,
        },
        /**
         * The mime type, can be changed if i.e server change the name.
         */
        mimeType: {
            type: String,
            required: false,
            default: undefined,
        },
        /**
         * The name of the file uploaded
         */
        originalMimeType: {
            type: String,
            required: false,
            default: undefined,
        },
        /**
         * If file name changed, this info will be displayed, placeholder %before% and %after% will be replaced with originalMimeType and mimeType
         */
        changedMimeTypeText: {
            type: String,
            required: false,
            default: undefined,
        },
    },
    computed: {
        isMimeTypeChanged(): boolean {
            return isSet(this.originalMimeType) && this.originalMimeType !== this.mimeType;
        },
        mimeTypeChangedText(): string {
            const originalMimeType = (this.originalMimeType || "").split("/").pop() || "";
            const currentMimeType = (this.mimeType || "").split("/").pop() || "";
            const localChangedMimeTypeText =
                this.changedMimeTypeText ??
                this.$t("fkui.file-item.changed-mime-type.text", "(%before% ändrad till %after%)");
            return localChangedMimeTypeText.replace("%before%", originalMimeType).replace("%after%", currentMimeType);
        },
        iconName(): IconName {
            const { mimeType } = this;

            /* use default icon when mimetype is empty/missing */
            if (!mimeType) {
                return DEFAULT_ICON;
            }

            /* search literal match */
            if (iconMap[mimeType]) {
                return iconMap[mimeType];
            }

            /* search with subtype wildcard, e.g. image/* */
            const p = mimeType.split("/", 1);
            const wildcard = [p[0], "*"].join("/");
            if (iconMap[wildcard]) {
                return iconMap[wildcard];
            }

            /* default to file */
            return DEFAULT_ICON;
        },
    },
});
</script>

<template>
    <div class="file-item">
        <div class="file-item__row">
            <a :id="id" class="file-item__file-open" v-bind="$attrs">
                <div class="icon-stack button__icon icon-stack--new-window">
                    <f-icon name="new-window"></f-icon>
                    <f-icon :name="iconName"></f-icon>
                </div>
                <span class="file-item__file-name">{{ fileName }}</span>
                <span class="sr-only">&nbsp;{{ $t("fkui.file-item.file-open", "öppnas i nytt fönster") }}</span>
            </a>

            <!--
            @slot Slot for any content that should be placed beside the file icon and file name. If several elements is placed, they will be spaced evenly, for more information see https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content.
            -->
            <slot name="row"></slot>
        </div>
        <!-- @slot Slot for any type of content. Elements placed in this slot will be position after the file icon and file name. -->
        <slot></slot>
        <div v-if="isMimeTypeChanged" class="file-item__change-info">{{ mimeTypeChangedText }}</div>
        <hr class="file-item__separator" />
    </div>
</template>
