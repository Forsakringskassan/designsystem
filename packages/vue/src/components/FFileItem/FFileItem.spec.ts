import { shallowMount, VueWrapper } from "@vue/test-utils";
import { TranslationPlugin } from "../../plugins";
import FFileItem from "./FFileItem.vue";

function createWrapper({ slots = {}, attrs = {}, props = {} }): VueWrapper {
    return shallowMount(FFileItem, {
        slots: slots,
        attrs: { ...attrs },
        props: {
            id: "123",
            fileName: "foo.bar",
            mimeType: "application/pdf",
            originalMimeType: "application/pdf",
            ...props,
        },
        global: {
            plugins: [TranslationPlugin],
        },
    });
}

describe("FileItem", () => {
    it("should match snapshot with slots", () => {
        const wrapper = createWrapper({
            slots: {
                row: "file item goes here",
                default: "progress is 80%",
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: "disabled",
            },
        });
        const aTag = wrapper.get(".file-item__file-open");
        expect(aTag.attributes("disabled")).toBeDefined();
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get(".file-item__file-open");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });

    it("should show info when server changed type", async () => {
        expect.assertions(1);
        const wrapper = createWrapper({
            props: { originalMimeType: "image/png" },
        });
        const element = wrapper.get(".file-item__change-info");
        expect(element.text()).toContain("(png Ã¤ndrad till pdf)");
    });

    it("should show custom info when server changed type", async () => {
        expect.assertions(1);
        const wrapper = createWrapper({
            props: {
                originalMimeType: "image/png",
                changedMimeTypeText: "(%before% changed to %after%)",
            },
        });
        const element = wrapper.get(".file-item__change-info");
        expect(element.text()).toContain("(png changed to pdf)");
    });

    it("should match snapshots without slots", () => {
        const wrapper = createWrapper({});
        expect(wrapper.element).toMatchSnapshot();
    });
});

it.each`
    mimeType                                                                     | expected
    ${"application/msword"}                                                      | ${"doc"}
    ${"application/octet-stream"}                                                | ${"file"}
    ${"application/pdf"}                                                         | ${"pdf"}
    ${"application/vnd.openxmlformats-officedocument.wordprocessingml.document"} | ${"doc"}
    ${"application/vnd.openxmlformats-officedocument.wordprocessingml.template"} | ${"doc"}
    ${"image/bmp"}                                                               | ${"pic"}
    ${"image/gif"}                                                               | ${"pic"}
    ${"image/jpeg"}                                                              | ${"pic"}
    ${"image/png"}                                                               | ${"pic"}
    ${"image/tiff"}                                                              | ${"pic"}
    ${"text/plain"}                                                              | ${"file"}
    ${"malformed-mimetype"}                                                      | ${"file"}
    ${"malformed/mime/type"}                                                     | ${"file"}
    ${""}                                                                        | ${"file"}
    ${undefined}                                                                 | ${"file"}
    ${null}                                                                      | ${"file"}
`("should have correct icon for $mimeType", ({ mimeType, expected }) => {
    const wrapper = createWrapper({
        props: {
            mimeType,
        },
    });
    const icon = wrapper.get("f-icon-stub").element;
    expect(icon.getAttribute("name")).toEqual(expected);
});
