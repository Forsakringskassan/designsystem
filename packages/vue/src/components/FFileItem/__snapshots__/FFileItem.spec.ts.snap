// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileItem should match snapshot with slots 1`] = `
<div
  class="file-item"
>
  <div
    class="file-item__row"
  >
    <a
      class="file-item__file-open"
      id="123"
    >
      <f-icon-stub
        library="f"
        name="pdf"
      />
      <span
        class="file-item__file-name"
      >
        foo.bar
      </span>
      <span
        class="sr-only"
      >
         öppnas i nytt fönster
      </span>
    </a>
    
    file item goes here
    
  </div>
  
  progress is 80%
  
  <!--v-if-->
  <hr
    class="file-item__separator"
  />
</div>
`;

exports[`FileItem should match snapshots without slots 1`] = `
<div
  class="file-item"
>
  <div
    class="file-item__row"
  >
    <a
      class="file-item__file-open"
      id="123"
    >
      <f-icon-stub
        library="f"
        name="pdf"
      />
      <span
        class="file-item__file-name"
      >
        foo.bar
      </span>
      <span
        class="sr-only"
      >
         öppnas i nytt fönster
      </span>
    </a>
    
    
  </div>
  
  
  <!--v-if-->
  <hr
    class="file-item__separator"
  />
</div>
`;

exports[`FileItem should pass attributes 1`] = `
<div
  class="file-item"
>
  <div
    class="file-item__row"
  >
    <a
      class="file-item__file-open"
      disabled="disabled"
      id="123"
    >
      <f-icon-stub
        library="f"
        name="pdf"
      />
      <span
        class="file-item__file-name"
      >
        foo.bar
      </span>
      <span
        class="sr-only"
      >
         öppnas i nytt fönster
      </span>
    </a>
    
    
  </div>
  
  
  <!--v-if-->
  <hr
    class="file-item__separator"
  />
</div>
`;
